File: scripts/security/scan-compromised.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const os = require('os');
let semver;
try {
  semver = require('semver');
} catch (e) {
  console.error('Missing dependency: semver. Install it with: npm i -D semver');
  process.exit(2);
}

function readJson(p) {
  return JSON.parse(fs.readFileSync(p, 'utf8'));
}

function loadSpecs(arg) {
  if (!arg) return [];
  if (arg.endsWith('.json')) return readJson(arg);
  return arg.split(',').map(s => {
    const at = s.lastIndexOf('@');
    if (at > 0) return { name: s.slice(0, at), range: s.slice(at + 1) };
    return { name: s, range: '*' };
  });
}

function collectInstalledFromLock(lock) {
  const map = new Map();
  const entries = [];

  if (lock.packages) {
    for (const [pkgPath, info] of Object.entries(lock.packages)) {
      if (!info || !info.version) continue;
      if (!pkgPath || !pkgPath.startsWith('node_modules/')) continue;
      const name = pkgPath.replace(/^node_modules\//, '').split('node_modules/').pop();
      const version = info.version;
      entries.push({ name, version, from: pkgPath });
      if (!map.has(name)) map.set(name, new Set());
      map.get(name).add(version);
    }
  } else if (lock.dependencies) {
    function walk(depTree, parents = []) {
      for (const [name, info] of Object.entries(depTree)) {
        if (!info || !info.version) continue;
        const version = info.version;
        entries.push({ name, version, from: [...parents, name].join(' > ') });
        if (!map.has(name)) map.set(name, new Set());
        map.get(name).add(version);
        if (info.dependencies) walk(info.dependencies, [...parents, name]);
      }
    }
    walk(lock.dependencies);
  }

  return { entries, map };
}

function classifyDirectness(pkgJson, name) {
  const direct =
    (pkgJson.dependencies && Object.prototype.hasOwnProperty.call(pkgJson.dependencies, name)) ||
    (pkgJson.devDependencies && Object.prototype.hasOwnProperty.call(pkgJson.devDependencies, name)) ||
    (pkgJson.optionalDependencies && Object.prototype.hasOwnProperty.call(pkgJson.optionalDependencies, name)) ||
    (pkgJson.peerDependencies && Object.prototype.hasOwnProperty.call(pkgJson.peerDependencies, name));
  return direct ? 'direct' : 'transitive';
}

function generateReport({ pkgJson, specs, hits, outputPath }) {
  const lines = [];
  const now = new Date().toISOString();

  lines.push('# Supply Chain Audit Report');
  lines.push('');
  lines.push(`Generated: ${now}`);
  lines.push('');
  lines.push('## Input (Compromised Packages)');
  specs.forEach(s => lines.push(`- ${s.name} (${s.range || '*'})`));
  if (!specs.length) lines.push('- (none provided)');
  lines.push('');

  lines.push('## Findings');
  if (!hits.length) {
    lines.push('- No matches found in installed dependency graph.');
  } else {
    for (const h of hits) {
      lines.push(`- ${h.name}@${h.version} — ${h.directness} dependency`);
      if (h.instances && h.instances.length) {
        h.instances.slice(0, 10).forEach(inst => lines.push(`  - via: ${inst.from}`));
        if (h.instances.length > 10) {
          lines.push(`  - ...and ${h.instances.length - 10} more instances`);
        }
      }
    }
  }
  lines.push('');

  lines.push('## Recommendations');
  if (hits.length) {
    lines.push('- Update to a safe version or replace the package.');
    lines.push('- If transitive, use "overrides" in package.json to pin safe versions.');
    lines.push('- Run: npm dedupe && npm install && npm audit');
    lines.push('- In CI, run with lifecycle scripts disabled: npm ci --ignore-scripts');
  } else {
    lines.push('- Keep dependencies pinned and lockfile committed.');
    lines.push('- Enable CI guardrails (see Measures below).');
  }
  lines.push('');

  lines.push('## Measures to avoid infection from upstream packages');
  lines.push('- Pin exact versions; commit package-lock.json; use "npm ci".');
  lines.push('- Enforce provenance/sig verification for npm publishes.');
  lines.push('- Disable lifecycle scripts in CI: npm ci --ignore-scripts.');
  lines.push('- Restrict registry to https://registry.npmjs.org/.');
  lines.push('- Enable Dependabot/Snyk/Socket alerts.');
  lines.push('- Use "overrides" to force safe transitive versions.');
  lines.push('');

  lines.push('## Measures to avoid infection from source');
  lines.push('- Code review and least-privilege on repo/CI tokens.');
  lines.push('- Secret scanning and pre-commit hooks.');
  lines.push('- Run dep scans on each PR.');
  lines.push('- Sandbox builds; avoid prod secrets on dev machines.');
  lines.push('- Verify integrity fields in lockfile.');
  lines.push('');

  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  fs.writeFileSync(outputPath, lines.join(os.EOL), 'utf8');
  return outputPath;
}

async function main() {
  const root = process.cwd();
  const pkgJsonPath = path.join(root, 'package.json');
  const lockPath = path.join(root, 'package-lock.json');

  if (!fs.existsSync(pkgJsonPath) || !fs.existsSync(lockPath)) {
    console.error('Missing package.json or package-lock.json in current directory.');
    process.exit(2);
  }

  const pkgJson = readJson(pkgJsonPath);
  const lock = readJson(lockPath);

  const arg = process.argv[2] || path.join(root, 'security', 'compromised.json');
  const specs = loadSpecs(arg);

  const { entries, map } = collectInstalledFromLock(lock);

  const hitList = [];
  for (const spec of specs) {
    const installedVersions = map.get(spec.name);
    if (!installedVersions) continue;
    const instances = entries.filter(e => e.name === spec.name);
    const matches = [];
    for (const v of installedVersions) {
      const coerced = semver.coerce(v) || v;
      const ok = spec.range && spec.range !== '*'
        ? semver.satisfies(coerced, spec.range, { includePrerelease: true, loose: true })
        : true;
      if (ok) matches.push(v);
    }
    if (matches.length) {
      const directness = classifyDirectness(pkgJson, spec.name);
      hitList.push({ name: spec.name, version: [...new Set(matches)].join(', '), directness, instances });
    }
  }

  const outDir = path.join(root, 'security');
  const p = generateReport({ pkgJson, specs, hits: hitList, outputPath: path.join(outDir, 'compromised-report.md') });
  console.log(`Report written to: ${p}`);
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: scripts/security/scan_compromised.py
#!/usr/bin/env python3
import argparse
import json
import os
from datetime import datetime


def load_json(path):
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


essential_dep_keys = [
    'dependencies', 'devDependencies', 'optionalDependencies', 'peerDependencies'
]


def classify_directness(pkg_json, name):
    for key in essential_dep_keys:
        deps = pkg_json.get(key, {})
        if name in deps:
            return 'direct'
    return 'transitive'


def collect_from_lock(lock):
    entries = []
    if isinstance(lock, dict) and 'packages' in lock:  # npm lockfile v2+
        for pkg_path, info in lock.get('packages', {}).items():
            if not info or not isinstance(info, dict):
                continue
            version = info.get('version')
            if not version:
                continue
            if not pkg_path or not pkg_path.startswith('node_modules/'):
                continue
            # normalize name for nested node_modules paths
            name = pkg_path.replace('node_modules/', '').split('node_modules/')[-1]
            entries.append({'name': name, 'version': version, 'from': pkg_path})
    elif isinstance(lock, dict) and 'dependencies' in lock:  # npm lockfile v1
        def walk(dep_tree, parents=None):
            parents = parents or []
            for name, info in dep_tree.items():
                if not info or not isinstance(info, dict):
                    continue
                version = info.get('version')
                if version:
                    entries.append({'name': name, 'version': version, 'from': ' > '.join(parents + [name])})
                if 'dependencies' in info:
                    walk(info['dependencies'], parents + [name])
        walk(lock['dependencies'])
    return entries


def version_matches(version: str, rng: str | None) -> bool:
    """Simple matcher for the ranges used in security/compromised.json.
    Supports:
      - exact version: "1.2.3" or "=1.2.3"
      - OR lists with "||" (e.g., "=4.1.1 || =4.1.2")
      - wildcard "*" or missing range => match any installed version
    """
    if not rng or rng.strip() == '*' or rng.strip() == '':
        return True
    parts = [p.strip() for p in rng.split('||')]
    for part in parts:
        if part.startswith('='):
            target = part[1:].strip()
        else:
            target = part
        if version == target:
            return True
    return False


def scan(root_dir: str, compromised_path: str) -> dict:
    pkg_json_path = os.path.join(root_dir, 'package.json')
    lock_path = os.path.join(root_dir, 'package-lock.json')
    if not (os.path.exists(pkg_json_path) and os.path.exists(lock_path)):
        raise SystemExit('Missing package.json or package-lock.json in ' + root_dir)

    pkg_json = load_json(pkg_json_path)
    lock = load_json(lock_path)
    specs = load_json(compromised_path)

    entries = collect_from_lock(lock)
    # name -> [instances]
    from collections import defaultdict
    instances_by_name = defaultdict(list)
    versions_by_name = defaultdict(set)
    for e in entries:
        instances_by_name[e['name']].append(e)
        versions_by_name[e['name']].add(e['version'])

    hits = []
    for spec in specs:
        name = spec.get('name')
        rng = spec.get('range')
        if not name:
            continue
        installed = versions_by_name.get(name)
        if not installed:
            continue
        matched_versions = sorted({v for v in installed if version_matches(v, rng)})
        if matched_versions:
            directness = classify_directness(pkg_json, name)
            hits.append({
                'name': name,
                'versions': matched_versions,
                'directness': directness,
                'instances': instances_by_name.get(name, [])
            })

    return {
        'root': root_dir,
        'specs': specs,
        'hits': hits,
    }


def write_report(root_dir: str, result: dict, out_rel: str = os.path.join('security', 'compromised-report-python.md')) -> str:
    out_path = os.path.join(root_dir, out_rel)
    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    lines = []
    lines.append('# Supply Chain Audit Report (Python)')
    lines.append('')
    lines.append(f'Generated: {datetime.utcnow().isoformat()}Z')
    lines.append('')
    lines.append('## Input (Compromised Packages)')
    for s in result['specs']:
        nm = s.get('name', '')
        rng = s.get('range', '*')
        lines.append(f'- {nm} ({rng})')
    lines.append('')

    lines.append('## Findings')
    if not result['hits']:
        lines.append('- No matches found in installed dependency graph.')
    else:
        for h in result['hits']:
            joined_versions = ', '.join(h['versions'])
            lines.append(f"- {h['name']}@{joined_versions} — {h['directness']} dependency")
            for inst in h['instances'][:10]:
                lines.append(f"  - via: {inst.get('from')}")
            extra = len(h['instances']) - 10
            if extra > 0:
                lines.append(f"  - ...and {extra} more instances")
    lines.append('')

    lines.append('## Recommendations')
    if result['hits']:
        lines.append('- Update to a safe version or replace the package.')
        lines.append('- If transitive, use "overrides" in package.json to pin safe versions.')
        lines.append('- Run: npm dedupe && npm install && npm audit')
        lines.append('- In CI, run with lifecycle scripts disabled: npm ci --ignore-scripts')
    else:
        lines.append('- Keep dependencies pinned and lockfile committed.')
        lines.append('- Enable CI guardrails (see Measures below).')
    lines.append('')

    lines.append('## Measures to avoid infection from upstream packages')
    lines.append('- Pin exact versions; commit package-lock.json; use "npm ci".')
    lines.append('- Enforce provenance/sig verification for npm publishes.')
    lines.append('- Disable lifecycle scripts in CI: npm ci --ignore-scripts.')
    lines.append('- Restrict registry to https://registry.npmjs.org/.')
    lines.append('- Enable Dependabot/Snyk/Socket alerts.')
    lines.append('- Use "overrides" to force safe transitive versions.')
    lines.append('')

    lines.append('## Measures to avoid infection from source')
    lines.append('- Code review and least-privilege on repo/CI tokens.')
    lines.append('- Secret scanning and pre-commit hooks.')
    lines.append('- Run dep scans on each PR.')
    lines.append('- Sandbox builds; avoid prod secrets on dev machines.')
    lines.append('- Verify integrity fields in lockfile.')

    with open(out_path, 'w', encoding='utf-8') as f:
        f.write('\n'.join(lines) + '\n')
    return out_path


def main():
    parser = argparse.ArgumentParser(description='Scan for compromised packages using package-lock.json')
    parser.add_argument('--root', default='.', help='Project root containing package.json and package-lock.json')
    parser.add_argument('--compromised', default=os.path.join('security', 'compromised.json'), help='Path to compromised.json list')
    parser.add_argument('--out', default=os.path.join('security', 'compromised-report-python.md'), help='Report output relative path under root')
    args = parser.parse_args()

    result = scan(args.root, args.compromised)
    path = write_report(args.root, result, args.out)
    print(f'Report written to: {path}')


if __name__ == '__main__':
    main()
